### PR TITLE
Hbond analysis update

### LINKED ISSUE #3933: modernize HBA to use AG as objects internally instead of selection strings
## Is your feature request related to a problem? ##
modernize HBA (HydrogenBondAnalysis) to use AG (AtomGroups) as objects internally instead of selection strings to avoid the need for HBA analysis to discern the relevant atoms.

## Describe the solution you'd like ##
AG as objects internally instead of selection strings 

## Additional context ##
See PR Hbond update #3848 (in particular https://github.com/MDAnalysis/mdanalysis/pull/3848#discussion_r1014548556)

## Guidance for new contributors
This is a reasonably complicated issue to solve for a beginner. We recommend you first learn about the problem. Ask questions here. The more specific your question, then better we can help.

As the first steps you should:

- Familiarize yourself with [AtomGroups](https://docs.mdanalysis.org/stable/documentation_pages/core/groups.html#MDAnalysis.core.groups.AtomGroup) and [UpdatingAtomGroups](https://docs.mdanalysis.org/stable/documentation_pages/core/groups.html#MDAnalysis.core.groups.UpdatingAtomGroup). Understand the difference between the two.
- Understand the difference between *using [selection strings](https://docs.mdanalysis.org/stable/documentation_pages/selections.html)* to describe groups of atoms and using AtomGroups for the same purpose.
- Read the code for [HydrogenBondAnalysis](https://docs.mdanalysis.org/stable/documentation_pages/analysis/hydrogenbonds.html) in [package/MDAnalysis/analysis/hydrogenbonds/hbond_analysis.py](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/analysis/hydrogenbonds/hbond_analysis.py). Be able to identify the pieces of code that use selection strings.

EDITS
- made issue clearer for new contributors (@orbeckst  )

### PR DESCRIPTION
**Pull request by yickysan**
Fixes #3933

Changes made in this Pull Request:
 - Changed the guess_hydrogens, guess_donors and guess_acceptors methods to use `ag.atoms` instead of `ag.select.atoms("all")` if a selection string isn't given.

PR Checklist
------------
 - [ ] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [ ] Issue raised/referenced?

----
:books: Documentation preview :books:: https://readthedocs-preview--4107.org.readthedocs.build/en/4107/

### CODE REVIEW

**Review by RMeli — CHANGES_REQUESTED** [review]
Hi @yickysan, thanks for you contribution. It looks like in this PR you also have the changes that you did for PR #4090. Could you please remove/revert them?

---

I see you are correctly working on a specific branch for this fix (`yickysan:hbond-analysis-update`) which is great. You probably created this branch from `yickysan:numpy.testing-function-change`, instead of creating it from `develop` and that's likely why you have both changes in this branch.

**Comment by yickysan** [comment]
> Hi @yickysan, thanks for you contribution. It looks like in this PR you also have the changes that you did for PR #4090. Could you please remove/revert them?
> 
> I see you are correctly working on a specific branch for this fix (`yickysan:hbond-analysis-update`) which is great. You probably created this branch from `yickysan:numpy.testing-function-change`, instead of creating it from `develop` and that's likely why you have both changes in this branch.

Hello, 
Thanks for pointing it out to me. I have deleted the `numpy.testing-function-change-branch.`  leaving only the `hbond-analysis-update`.

**Comment by RMeli** [comment]
@yickysan as you can see from the [files changed](https://github.com/MDAnalysis/mdanalysis/pull/4107/files), deleting the `yickysan:numpy.testing-function-change` branch (or closing PR #4090) does not solve the issue. This is because the changes have been included in the `yickysan:hbond-analysis-update` branch as well (the likely explanation of how this happened is in the previous message).

You will need to revert the changes _in this branch_. You can probably use [`git revert`](https://git-scm.com/docs/git-revert) to revert the commit(s) related to the other branch. More specifically you need to revert the following commit (please double check!):
* 0d7bfd4b90626c1e43dc39a552493a78b79dc53c
* 9c116315d946b44a55be972e7ecb0409d6952bbd
* 4086defb873b279d9637cfbf79257e7a0f185ec0

(These are the commits that you also had in the other PR (see [MDAnalysis/pull/4090/commits](https://github.com/MDAnalysis/mdanalysis/pull/4090/commits) and the commit message appears to be related to the tests).

**Comment by yickysan** [comment]
> @yickysan as you can see from the [files changed](https://github.com/MDAnalysis/mdanalysis/pull/4107/files), deleting the `yickysan:numpy.testing-function-change` branch (or closing PR #4090) does not solve the issue. This is because the changes have been included in the `yickysan:hbond-analysis-update` branch as well (the likely explanation of how this happened is in the previous message).
> 
> You will need to revert the changes _in this branch_. You can probably use [`git revert`](https://git-scm.com/docs/git-revert) to revert the commit(s) related to the other branch. More specifically you need to revert the following commit (please double check!):
> 
> * [0d7bfd4](https://github.com/MDAnalysis/mdanalysis/commit/0d7bfd4b90626c1e43dc39a552493a78b79dc53c)
> * [9c11631](https://github.com/MDAnalysis/mdanalysis/commit/9c116315d946b44a55be972e7ecb0409d6952bbd)
> * [4086def](https://github.com/MDAnalysis/mdanalysis/commit/4086defb873b279d9637cfbf79257e7a0f185ec0)
> 
> (These are the commits that you also had in the other PR (see [MDAnalysis/pull/4090/commits](https://github.com/MDAnalysis/mdanalysis/pull/4090/commits) and the commit message appears to be related to the tests).

I am currently reverting these commits as well.

**Comment by yickysan** [comment]
> @yickysan as you can see from the [files changed](https://github.com/MDAnalysis/mdanalysis/pull/4107/files), deleting the `yickysan:numpy.testing-function-change` branch (or closing PR #4090) does not solve the issue. This is because the changes have been included in the `yickysan:hbond-analysis-update` branch as well (the likely explanation of how this happened is in the previous message).
> 
> You will need to revert the changes _in this branch_. You can probably use [`git revert`](https://git-scm.com/docs/git-revert) to revert the commit(s) related to the other branch. More specifically you need to revert the following commit (please double check!):
> 
> * [0d7bfd4](https://github.com/MDAnalysis/mdanalysis/commit/0d7bfd4b90626c1e43dc39a552493a78b79dc53c)
> * [9c11631](https://github.com/MDAnalysis/mdanalysis/commit/9c116315d946b44a55be972e7ecb0409d6952bbd)
> * [4086def](https://github.com/MDAnalysis/mdanalysis/commit/4086defb873b279d9637cfbf79257e7a0f185ec0)
> 
> (These are the commits that you also had in the other PR (see [MDAnalysis/pull/4090/commits](https://github.com/MDAnalysis/mdanalysis/pull/4090/commits) and the commit message appears to be related to the tests).

Please I need your help. In an attempt to revert the commits I have accidentally deleted some files. Anyway you can help me through this?

> > @yickysan as you can see from the [files changed](https://github.com/MDAnalysis/mdanalysis/pull/4107/files), deleting the `yickysan:numpy.testing-function-change` branch (or closing PR #4090) does not solve the issue. This is because the changes have been included in the `yickysan:hbond-analysis-update` branch as well (the likely explanation of how this happened is in the previous message).
> > You will need to revert the changes _in this branch_. You can probably use [`git revert`](https://git-scm.com/docs/git-revert) to revert the commit(s) related to the other branch. More specifically you need to revert the following commit (please double check!):
> > 
> > * [0d7bfd4](https://github.com/MDAnalysis/mdanalysis/commit/0d7bfd4b90626c1e43dc39a552493a78b79dc53c)
> > * [9c11631](https://github.com/MDAnalysis/mdanalysis/commit/9c116315d946b44a55be972e7ecb0409d6952bbd)
> > * [4086def](https://github.com/MDAnalysis/mdanalysis/commit/4086defb873b279d9637cfbf79257e7a0f185ec0)
> > 
> > (These are the commits that you also had in the other PR (see [MDAnalysis/pull/4090/commits](https://github.com/MDAnalysis/mdanalysis/pull/4090/commits) and the commit message appears to be related to the tests).
> 
> I am currently reverting these commits as well.

Hello is it possible to close this pull requests while I make another one? I think in my attempt to revert the other commits I might have broken something.

> > * [](https://github.com/MDAnalysis/mdanalysis/commit/9c116315d946b44a55be972e7ecb0409d6952bbd)
> > * [4086def](https://github.com/MDAnalysis/mdanalysis/commit/4086defb873b279d9637cfbf79257e7a0f185ec0)
> > 
> > (These are the commits that you also had in the other PR (see [MDAnalysis/pull/4090/commits](https://github.com/MDAnalysis/mdanalysis/pull/4090/commits) and the commit message appears to be related to the tests).
> 
> I am currently re

**Comment by RMeli** [comment]
How did you delete the file? Were you using `git revert`? Since you deleted the whole `testsuite/MDAnalysisTests/analysis/test_align.py` file, I think the simplest fix here would be to download such file from the `develop` branch in the MDAnalysis repo and add it back to your repo. This way, you should have the file back, and also revert your commits regarding the file.

Don't be discouraged by these difficulties with `git`. They are quite normal and get better with practice: [xkcd.com/1597](https://xkcd.com/1597/) =)

**Comment by yickysan** [comment]
> How did you delete the file? Were you using `git revert`? Since you deleted the whole `testsuite/MDAnalysisTests/analysis/test_align.py` file, I think the simplest fix here would be to download such file from the `develop` branch in the MDAnalysis repo and add it back to your repo. This way, you should have the file back, and also revert your commits regarding the file.
> 
> Don't be discouraged by these difficulties with `git`. They are quite normal and get better with practice: [xkcd.com/1597](https://xkcd.com/1597/) =)

I think I must have accidentally deleted the file trying to handle a merge conflict. I used git reset and not revert. I have redownloaded the file and added it back to my local repo.

**Comment by RMeli** [comment]
Make sure to commit and push, so that the file is restored in this PR as well. If everything goes well, it should disappear from the "files changed" tab.

You can read here about the difference between reset and revert: https://git-scm.com/docs/git#_reset_restore_and_revert The latter is safer because it does not change the commit history.

**Comment by RMeli** [comment]
@yickysan Looks like you removed all the changes, including the ones related to this PR? There is nothing to review at the moment. (I'm on mobile, apologies if I missed something ing.)

**Comment by RMeli** [comment]
You can use "git status" and "git diff" locally to check what's going and in which state is your (local) repo.

**Comment by yickysan** [comment]
> @yickysan Looks like you removed all the changes, including the ones related to this PR? There is nothing to review at the moment. (I'm on mobile, apologies if I missed something ing.)

Yes the revert I made removed the file related to this pull request but I am adding it back right now.

**Comment by yickysan** [comment]
> You can use "git status" and "git diff" locally to check what's going and in which state is your (local) repo.

Hello @RMeli I added back the commit related to this PR. It failed the codecov tests. Is there a way to fix this?

File: package/MDAnalysis/analysis/hydrogenbonds/hbond_analysis.py [hunk 1]
```diff
@@ -434,18 +434,29 @@ def guess_hydrogens(self,
...
-        hydrogens_ag = ag[
-            np.logical_and.reduce((
-                ag.masses < max_mass,
-                ag.charges > min_charge,
-                ag.masses > min_mass,
-            ))
-        ]
+        if select:
+
+            ag = self.u.select_atoms(select)
```
Comment (IAlibay) [inline package/MDAnalysis/analysis/hydrogenbonds/hbond_analysis.py:439]: Ratther than doing a completely new code branch for each option, why not just do this selection solely in an if/else and leaving everything else the same?

**Review by IAlibay — CHANGES_REQUESTED** [review]
I'm not sure I understand the intent behind these changes, how is this intended to change the usability of the hbond analysis class? Maybe if you can provide a quick walked through example of how a user might help review.

re coverage: this is happening because you don't have any tests that cover the new code branches that were introduced in this PR. To avoid them, you would need to add tests that try out the new input method you've introduced.

**Comment by yickysan** [comment]
> I'm not sure I understand the intent behind these changes, how is this intended to change the usability of the hbond analysis class? Maybe if you can provide a quick walked through example of how a user might help review.
> 
> re coverage: this is happening because you don't have any tests that cover the new code branches that were introduced in this PR. To avoid them, you would need to add tests that try out the new input method you've introduced.

So it doesn't change the usability of the hbondanalysis class.  It just changes the `guess_donors`, `guess_hydrogens` methods, etc to call `u.atoms` if a selection string isn't given instead of calling `u.select_atoms("all")`

**Pull request closed without merging**